### PR TITLE
supported upsert points batch

### PR DESCRIPTION
**Pull request by Nik4Sun**

### CODE REVIEW

File: tests/Aer.QdrantClient.Tests/TestClasses/HttpClientTests/PointsCrudTests.UpsertDelete.cs [hunk 1]
```diff
@@ -645,6 +1152,93 @@ var upsertPointsResult
             readPointPayload.Text.Should().Be(expectedPointPayload.Text);
         }
     }
+    
+        [Test]
```
Comment (TonySkorik) [inline tests/Aer.QdrantClient.Tests/TestClasses/HttpClientTests/PointsCrudTests.UpsertDelete.cs]: Unnecessary indent

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/ObjectPayloadEnumerableJsonConverter.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Aer.QdrantClient.Http.Infrastructure.Json.Converters;
+
+internal sealed class ObjectPayloadEnumerableJsonConverter : JsonConverter<IEnumerable<object>>
```
Comment (TonySkorik) [inline src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/ObjectPayloadEnumerableJsonConverter.cs:6]: The names of the file and the class are mismatched.
I believe that was initially my mistake. Let's also rename the `UpsertPointPayloadJsonConverter.cs` to `ObjectPayloadJsonConverter.cs`

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/UpsertPointPayloadEnumerableJsonConverter.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Aer.QdrantClient.Http.Infrastructure.Json.Converters;
+
+internal sealed class ObjectPayloadEnumerableJsonConverter : JsonConverter<IEnumerable<object>>
+{
+    private static readonly JsonSerializerOptions _serializerOptions =
+        JsonSerializerConstants.CreateSerializerOptions(new ObjectPayloadJsonConverter());
+
+    public override IEnumerable<object> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
+        throw new NotSupportedException("Reading upsert point payload object instances is not supported");
```
Comment (TonySkorik) [inline src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/UpsertPointPayloadEnumerableJsonConverter.cs]: Since this converter doesn't know anything about upsert points batch let's rephrase this message to 
`Reading enumerable point payload instances is not supported`.
Let's do the same in `ObjectPayloadJsonConverter`.

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+namespace Aer.QdrantClient.Http.Infrastructure.Json.Converters;
+
+internal sealed class VectorsBatchJsonConverter : JsonConverter<UpsertPointsRequest.UpsertPointsBatch.VectorsBatch>
+{
+   private static readonly JsonSerializerOptions _serializerOptions =
+        JsonSerializerConstants.CreateSerializerOptions(new VectorJsonConverter());
+
+    public override UpsertPointsRequest.UpsertPointsBatch.VectorsBatch Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        throw new NotSupportedException("Reading upsert vectors' batch is not supported");
```
Comment (TonySkorik) [inline src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs]: Again this converter is not strictly related to upsert points so let's rephrase this to `Reading vectors batch is not supported`

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            writer.WriteStartArray();
+
+            foreach (var vector in value.Vectors)
+            {
+                JsonSerializer.Serialize(writer, vector, _serializerOptions);
+            }
+
+            writer.WriteEndArray();
+        }
+        else if (value.NamedVectors?.Any() == true)
```
Comment (TonySkorik) [inline src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs]: Let's avoid calling Linq on non-enumerable types and use pattern matching instead : `value.NamedVectors is { Count: > 0 }`. This will also test for null.

File: src/Aer.QdrantClient.Http/Models/Requests/Public/UpsertPointsRequest.cs [hunk 6]
```diff
@@ -54,11 +54,90 @@ public UpsertPoint(
             Payload = payload;
         }
     }
+    
+    /// <summary>
+    /// Represents a points' batch to upsert to qdrant.
+    /// </summary>
+    public sealed class UpsertPointsBatch
```
Comment (TonySkorik) [inline src/Aer.QdrantClient.Http/Models/Requests/Public/UpsertPointsRequest.cs]: Since we are using this type in both `UpsertPointsRequest` and `BatchUpdatePointsRequest`, lets move it to a separate file under `Aer.QdrantClient.Http.Models.Requests.Public.Shared`.

**Review by TonySkorik — CHANGES_REQUESTED** [review]

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+         "payloads": [
+           { "test": "a" },
+           { "test": "b" }
+         ]
+       }
+     }
+   */
+    public override void Write(Utf8JsonWriter writer, UpsertPointsRequest.UpsertPointsBatch.VectorsBatch value, JsonSerializerOptions options)
+    {
+        if (HasVectors(value.Vectors))
```
Comment (TonySkorik) [inline src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs:58]: I took the liberty of optimising this count calculation a bit by using `TryGetNonEnumeratedCount`.
Because it's not available under netstandard we are using `NetstandardPolyfill` with inbuilt conditional compilation.

**Review by TonySkorik — CHANGES_REQUESTED** [review]

File: tests/Aer.QdrantClient.Tests/TestClasses/HttpClientTests/PointsCrudTests.UpsertDelete.cs (hunk 1 above)
Comment (Nik4Sun) [reply tests/Aer.QdrantClient.Tests/TestClasses/HttpClientTests/PointsCrudTests.UpsertDelete.cs]: Removed

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/ObjectPayloadEnumerableJsonConverter.cs (hunk 2 above)
Comment (Nik4Sun) [reply src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/ObjectPayloadEnumerableJsonConverter.cs:6]: Renamed

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/UpsertPointPayloadEnumerableJsonConverter.cs (hunk 3 above)
Comment (Nik4Sun) [reply src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/UpsertPointPayloadEnumerableJsonConverter.cs]: Updated

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs (hunk 4 above)
Comment (Nik4Sun) [reply src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs]: Updated

File: src/Aer.QdrantClient.Http/Models/Requests/Public/UpsertPointsRequest.cs (hunk 6 above)
Comment (Nik4Sun) [reply src/Aer.QdrantClient.Http/Models/Requests/Public/UpsertPointsRequest.cs]: Moved

File: src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs (hunk 5 above)
Comment (Nik4Sun) [reply src/Aer.QdrantClient.Http/Infrastructure/Json/Converters/VectorsBatchJsonConverter.cs]: Done
